src/sqlfluff/core/parser/segments/base.py

```diff
@@ -885,25 +885,24 @@ def select_children(
                 buff.append(seg)
         return buff
 
-    def recursive_crawl(self, *seg_type, recurse_into=True):
-        """Recursively crawl for segments of a given type.
-
-        Args:
-            seg_type: :obj:`str`: one or more type of segment
-                to look for.
-            recurse_into: :obj:`bool`: When an element of type "seg_type" is
-                found, whether to recurse into it.
-        """
+    def recursive_crawl(
+        self, *seg_type: str, recurse_into: bool = True, all_descendants: bool = False
+    ):
+        """Recursively crawl for segments of a given type (or all descendants)."""
         # Check this segment
-        if self.is_type(*seg_type):
+        if all_descendants or self.is_type(*seg_type):
             match = True
             yield self
```

Comment:
This feels weird.

So If I say `recursive_crawl(seg_type="keywords", recurse_into=true, all_descendants=true)` it will return all descendents, even those that are not `keywords`?

Feels to me like this is a completely different (though admittedly similar!) function and should either provide a type, or `all_descendants` but not both.

Would a type of value `ALL` or similar negate the need for the new (confusing and sometimes contradictory) param?

Reply:
Or make the type optional, and if it is not given, then that implies you want all descendants.

Reply:
Yes, it's weird. I was wondering about this myself, and I did this more out of laziness than feeling sure it was good design. I think providing a separate function will be safer/more consistent than treating "empty" as all or defining a special "ALL" value.

Reply:
Ok, I created a new, separate `recursive_crawl_all()` function.